torchinductor/lowering.py

```diff
@@ -2526,6 +2543,52 @@ def copy_(dst, src, non_blocking=False):
     return mutate_to(dst, src)
 
 
+@make_pointwise
+def floordiv(a, b):
+    return ops.floordiv(a, b)
+
+
+@make_pointwise
+def truncdiv(a, b):
+    return ops.truncdiv(a, b)
+
+
+@register_lowering(aten.div.Tensor_mode)
+def div_mode(a, b, rounding_mode=None):
+    both_integer = is_integer_type(a) and is_integer_type(b)
+    both_boolean = is_boolean_type(a) and is_boolean_type(b)
+    on_triton = is_triton(a) or is_triton(b)
+
+    # floordiv and truncdiv need special handling for integer tensors on Triton,
+    # see the discussion at https://github.com/openai/triton/issues/605
+    if rounding_mode == "floor":
+        assert not both_boolean, "floordiv operands can not be boolean at the same time"
+        return floordiv(a, b) if (both_integer and on_triton) else floor(div(a, b))
+    if rounding_mode == "trunc":
+        assert not both_boolean, "truncdiv operands can not be boolean at the same time"
+        return truncdiv(a, b) if (both_integer and on_triton) else trunc(div(a, b))
+    return div(a, b)
+
+
+@register_lowering(aten.div)
+def div(a, b):
+    def fn(*args):
+        return ops.div(*args)
+
+    both_integer = is_integer_type(a) and is_integer_type(b)
+    # truediv produces a float tensor even if both operands are integer types
+    dtype = (
+        torch.get_default_dtype()
+        if both_integer
+        else torch.promote_types(a.get_dtype(), b.get_dtype())
```

Comment:
Better use `torch.result_type(a, b)` here, that would solve test errors and also get correct result when one of the inputs is 0-dim tensor (promote_types wouldn't)